### PR TITLE
ASH support

### PR DESCRIPTION
**Pull request by alexander-demichev**
This PR adds:
- ability to create stackhub client by looking at infrastructure object.
- field to provider spec with availability set name
- stackhub services that use stackhub API.
- a wrapper that avoids importing specific api in actuator.

### CODE REVIEW

File: pkg/cloud/azure/services/networkinterfaces/networkinterfaces_stack.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/applicationsecuritygroups"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/internalloadbalancers"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/publicips"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/publicloadbalancers"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/securitygroups"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/subnets"
+)
+
+// StackHubSpec specification for routetable
+type StackHubSpec struct {
```
Comment (JoelSpeed) [inline pkg/cloud/azure/services/networkinterfaces/networkinterfaces_stack.go:38]: This wrapper is complicated, and a prime target for aiming some deduplication at

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -228,8 +227,14 @@ func (s *Reconciler) Update(ctx context.Context) error {
 						continue
 					}
 
-					ip, ok := publicIPInterface.(network.PublicIPAddress)
-					if !ok {
+					// ip, ok := publicIPInterface.(network.PublicIPAddress)
```
Comment (lobziik) [inline pkg/cloud/azure/actuators/machine/reconciler.go:230]: Pls remove it, or leave some meaningful comment, why it is there

File: pkg/cloud/azure/services/subnets/subnets.go [hunk 3]
```diff
@@ -83,7 +83,7 @@ func (s *Service) CreateOrUpdate(ctx context.Context, spec azure.Spec) error {
 	}
 	nsg, ok := nsgInterface.(network.SecurityGroup)
 	if !ok {
-		return errors.New("error getting network security group")
```
Comment (lobziik) [inline pkg/cloud/azure/services/subnets/subnets.go:86]: ?

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 4]
```diff
@@ -126,7 +125,7 @@ func (s *Reconciler) CreateMachine(ctx context.Context) error {
 
 // Update updates machine if and only if machine exists, handled by cluster-api
 func (s *Reconciler) Update(ctx context.Context) error {
-	vmSpec := &virtualmachines.Spec{
+	vmSpec := &virtualmachines.StackHubSpec{
```
[the remark this answers is not in the document]
Comment (alexander-demichev) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: this is outdated

**Comment by lobziik** [comment]
Couple of questions:
- would custom CA without proxy settings being handled?
- How it was tested? Would be nice to run our e2e suite against ASH installation.

**Comment by alexander-demichev** [comment]
@lobziik CA without proxy settings should be double-checked by QE, if that doesn't work I believe that the fix should be made in MAPI repository.
I did some manual testing of this code, I'm running full e2e will take all ASH capacity because of how many machines we use.

**Comment by freedge** [comment]
btw I tested this manually too, works pretty nice :)  can't wait for the delivery \o/

**Pull request merged**